File: ruby sketch/spec/game_spec.rb
require './lib/game'
describe Game do
  puts "GAME 3"
  it 'can correctly increment its score for ordinary rolls' do
    game = Game.new([5,4,3,2,3,2,4,1])
    expect(game.score_for_frame(1)).to eq(9)
    expect(game.score_for_frame(2)).to eq(14)
    expect(game.score_for_frame(3)).to eq(19)
    expect(game.score_for_frame(4)).to eq(24)
  end

  it 'can correctly increment its score if there is a spare' do
    game = Game.new([5,5,3,2,3,2,4,1])
    expect(game.score_for_frame(1)).to eq(13)
    expect(game.score_for_frame(2)).to eq(18)
    expect(game.score_for_frame(3)).to eq(23)
    expect(game.score_for_frame(4)).to eq(28)
  end

  it 'can correctly increment its score if there is a strike' do
    game = Game.new([10,3,2,10,4,1])
    expect(game.score_for_frame(1)).to eq(15)
    expect(game.score_for_frame(2)).to eq(20)
    expect(game.score_for_frame(3)).to eq(35)
    expect(game.score_for_frame(4)).to eq(40)
  end
    it 'can correctly increment its score if there is a strike' do
    game = Game.new([10,3,2,10,4,1])
    expect(game.score_for_frame(1)).to eq(15)
    expect(game.score_for_frame(2)).to eq(20)
    expect(game.score_for_frame(3)).to eq(35)
    expect(game.score_for_frame(4)).to eq(40)
  end

  it 'can correctly increment its score if there a 2 strikes in a row' do
    game = Game.new([10,10,5,4,6,2])
    expect(game.score_for_frame(1)).to eq(25)
    expect(game.score_for_frame(2)).to eq(44)
    expect(game.score_for_frame(3)).to eq(53)
    expect(game.score_for_frame(4)).to eq(61)
  end
  it 'can correctly increment its score if there a 3 strikes in a row' do
    game = Game.new([10,10,10,6,2])
    expect(game.score_for_frame(1)).to eq(30)
    expect(game.score_for_frame(2)).to eq(56)
    expect(game.score_for_frame(3)).to eq(74)
    expect(game.score_for_frame(4)).to eq(82)
  end

  it 'can handle 2 spares in a row' do
    game = Game.new([5,5,4,6,7,1])
    expect(game.score_for_frame(1)).to eq(14)
    expect(game.score_for_frame(2)).to eq(31)
    expect(game.score_for_frame(3)).to eq(39)
  end

  it 'can handle a strike and a spare' do
    game = Game.new([10,4,6,7,1])
    expect(game.score_for_frame(1)).to eq(20)
    expect(game.score_for_frame(2)).to eq(37)
    expect(game.score_for_frame(3)).to eq(45)
  end

  it 'can handle 12 strikes' do
    game = Game.new([10,10,10,10,10,10,10,10,10,10,10,10])
    expect(game.score_for_frame(10)).to eq(300)
  end
end

File: jasmine-standalone-2.8.0/spec/GameSpec.js
'use strict';

describe("Game", function() {

  var game;

  beforeEach(function() {
    game = new Game();
  })

  it("should initialise with a score of zero", function() {
    expect(game.score).toEqual(0);
  })
  it("should add the score from the frame to the game score", function() {
    game.bowl(3)
    game.bowl(4)
    expect(game.score).toEqual(7)
  })
  it('should start a new frame on the third roll (no strike)', function() {
    game.bowl(2)
    game.bowl(6)
    game.bowl(2)
    expect(game.frameNumber).toEqual(2)
    expect(game.currentFrame.score).toEqual(2)
  })
  it('should start a new frame after a strike', function() {
    game.bowl(10)
    game.bowl(6)
    expect(game.frameNumber).toEqual(2)
    expect(game.currentFrame.score).toEqual(6)
  })
  it("should add the frame to the frames array once it begins", function() {
    game.bowl(2)
    game.bowl(6)
    game.bowl(2)
    expect(game.frames.length).toEqual(2)
  })
  describe("Strikes and spares", function() {
    it("should add the next two bowled scores to a frame with a strike", function() {
      game.bowl(10)
      game.bowl(6)
      game.bowl(3)
      expect(game.score).toEqual(28)
    })
    it("should add the next bowled score to a frame with a spare", function() {
      game.bowl(5)
      game.bowl(5)
      game.bowl(3)
      game.bowl(2)
      expect(game.score).toEqual(18)
    })
    it("should move to the next frame when a strike is scored", function() {
      game.bowl(10)
      expect(game.frameNumber).toEqual(2)
      game.bowl(10)
      expect(game.frameNumber).toEqual(3)
    })
    it("should calculate scores with consecutive strikes", function() {
      game.bowl(10)
      game.bowl(10)
      game.bowl(10)
      game.bowl(10)
      expect(game.score).toEqual(90)
    })
    it("should add ten to a spare score when followed by a strike", function() {
      game.bowl(5)
      game.bowl(5)
      game.bowl(10)
      expect(game.score).toEqual(30)
    })
    it("should count a frame containing rolls of 0 and 10 as a spare", function() {
      game.bowl(0)
      game.bowl(10)
      game.bowl(0)
      game.bowl(2)
      expect(game.score).toEqual(12)
    })
    it("should calculate a mix of spares, strikes and consecutive strikes", function() {
      game.bowl(5)
      game.bowl(5)
      game.bowl(10)
      game.bowl(10)
      game.bowl(10)
      game.bowl(10)
      game.bowl(3)
      game.bowl(2)
      game.bowl(10)
      game.bowl(6)
      game.bowl(1)
      expect(game.score).toEqual(147)
    })
  })
  describe("Final frame", function() {
    it("should allow three rolls in the final frame if the first two are a spare", function() {
      game.frameNumber = 9
      game.bowl(10)
      game.bowl(5)
      game.bowl(5)
      expect(game.currentFrame.isFrameOver).toEqual(false)
    })
    it("should allow three rolls in the final frame if the first is a strike", function() {
      game.frameNumber = 9
      game.bowl(10)
      game.bowl(10)
      game.bowl(10)
      expect(game.frameNumber).toEqual(10)
      expect(game.currentFrame.isFrameOver).toEqual(false)
    })
    it("shouldn't allow the second roll to take the first two rolls over 10 unless the first was a strike", function() {
      game.frameNumber = 9
      game.bowl(1)
      game.bowl(1)
      game.bowl(9)
      game.bowl(2)
      expect(game.score).toEqual(11)
    })
    it("shouldn't allow the second and third rolls to total more than 10 unless the second was a strike", function() {
      game.frameNumber = 9
      game.bowl(1)
      game.bowl(1)
      game.bowl(10)
      game.bowl(9)
      game.bowl(2)
      expect(game.score).toEqual(21)
      expect(game.currentFrame.isFrameOver).toEqual(false)
    })
  })
  it("should correctly score a perfect game", function() {
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    expect(game.score).toEqual(300)
  })
  it("shouldn't allow players to bowl once the game is finished", function() {
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    game.bowl(10)
    expect(game.score).toEqual(300)
  })
  it("shouldn't allow a second roll of the frame to take the total frame score past 10", function() {
    game.bowl(9)
    game.bowl(2)
    expect(game.score).toEqual(9)
  })

})